mlflow/pyspark/ml/__init__.py

```diff
@@ -101,37 +111,115 @@ def _get_estimator_info_tags(estimator):
+        if is_tuning_estimator and k == 'estimator':
+            expanded_param_map[k] = v.uid
+            # skip log estimator's nested params because they will be
+            # logged in nested runs.
+            # TODO:
+            #  consider the nested case:
+            #  the tuned estimator is a pipeline or a composite estimator like OneVsRest
+            #  then nested estimators uid need logged, e.g. uids of stage in
+            #  pipeline.stages/ova.classifier
+            pass
+        elif is_tuning_estimator and k == 'estimatorParamMaps':
+            # this param will be saved as JSON format artifact.
+            pass
+        elif isinstance(v, Params):
             # handle the case param value type inherits `pyspark.ml.param.Params`
             # e.g. param like `OneVsRest.classifier`/`CrossValidator.estimator`
             expanded_param_map[k] = v.uid
             internal_param_map = _get_instance_param_map(v)
             for ik, iv in internal_param_map.items():
                 expanded_param_map[f"{v.uid}.{ik}"] = iv
-        elif k in ["estimator", "estimatorParamMaps"] and isinstance(
-            instance, (CrossValidator, TrainValidationSplit)
-        ):
-            # skip log estimator Param and its nested params because they will be
-            # logged in nested runs.
-            # TODO: Log `estimatorParamMaps` as JSON artifacts.
-            pass
         else:
             expanded_param_map[k] = v
 
     return expanded_param_map
 
 
+def _create_child_runs_for_parameter_search(parent_estimator, parent_model, parent_run, child_tags):
+    from pyspark.ml.tuning import CrossValidatorModel, TrainValidationSplitModel
+    from itertools import zip_longest
+
+    client = MlflowClient()
+    # Use the start time of the parent parameter search run as a rough estimate for the
+    # start time of child runs, since we cannot precisely determine when each point
+    # in the parameter search space was explored
+    child_run_start_time = parent_run.info.start_time
+    child_run_end_time = int(time.time() * 1000)
+
+    estimator_param_maps = parent_estimator.getEstimatorParamMaps()
+    tuning_estimator = parent_estimator.getEstimator()
+    metric_key = parent_estimator.getEvaluator().getMetricName()
+    if isinstance(parent_model, CrossValidatorModel):
+        metric_key = 'avg_' + metric_key
+        metrics = parent_model.avgMetrics
+    elif isinstance(parent_model, TrainValidationSplitModel):
+        metrics = parent_model.avgMetrics
+    else:
+        raise RuntimeError(f'Unknown model type {type(parent_model)}.')
+
+    for i in range(len(estimator_param_maps)):
+        estimator = tuning_estimator.copy(estimator_param_maps[i])
+        tags_to_log = dict(child_tags) if child_tags else {}
+        tags_to_log.update({MLFLOW_PARENT_RUN_ID: parent_run.info.run_id})
+        tags_to_log.update(_get_estimator_info_tags(estimator))
+        # TODO: add child run index into tag ?
```

Comment:
Shall we log index ?

Reply:
Is there any useful use case?

Reply:
Is the index helpful / valuable? If not, I don't think we should include it.

Reply:
@WeichenXu123 can we remove the TODO from line 223?